packages/features/ee/dsync/lib/handleGroupEvents.ts

```diff
@@ -0,0 +1,178 @@
+    where: {
+      directoryId: event.directory_id,
+      groupName: eventData.name,
+      orgId,
+    },
+    select: {
+      teamId: true,
+      team: {
+        include: {
+          parent: {
+            include: {
+              organizationSettings: true,
+            },
+          },
+          organizationSettings: true,
+        },
+      },
+      groupName: true,
+    },
+  });
+
+  if (!groupNames.length) {
+    return;
+  }
+
+  const org = await getTeamOrThrow(orgId);
+
+  // Check if the group member display property is an email
+  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+  const isEmail = emailRegex.test(eventData.raw.members[0].display);
+
+  let userEmails: string[] = [];
+
+  // TODO: Handle the case where display property is not an email
+
+  if (isEmail) {
+    userEmails = eventData.raw.members.map((member) => member.display);
+  }
+
+  // Find existing users
+  const users = await prisma.user.findMany({
+    where: {
+      email: {
+        in: userEmails,
+      },
+    },
+    select: {
+      id: true,
+      email: true,
+      username: true,
+      organizationId: true,
+      completedOnboarding: true,
+      identityProvider: true,
+      profiles: true,
+      locale: true,
+      profiles: {
+        select: {
+          organizationId: true,
+        },
+      },
+      teams: {
+        select: {
+          id: true,
+        },
+      },
+      password: {
+        select: {
+          hash: true,
+        },
+      },
+    },
+  });
+
+  const translation = await getTranslation("en", "common");
+
+  const newUserEmails = userEmails.filter((email) => !users.find((user) => user.email === email));
+  // For each team linked to the dsync group name provision members
+  for (const group of groupNames) {
+    if (newUserEmails.length) {
+      await createUsersAndConnectToOrg({
+        emailsToCreate: newUserEmails,
+        org,
+      });
+      await Promise.all(
+        newUserEmails.map((email) => {
+          return sendSignupToOrganizationEmail({
+            usernameOrEmail: email,
+            team: { ...group.team, metadata: teamMetadataSchema.parse(group.team.metadata) },
+            translation,
+            inviterName: org.name,
+            teamId: group.teamId,
+            isOrg: false,
+          });
+        })
+      );
+    }
+
+    // For existing users create membership for team and org if needed
+    await prisma.membership.createMany({
+      data: [
+        ...users
+          .map((user) => {
+            return [
+              {
+                userId: user.id,
+                teamId: group.teamId,
+                role: "MEMBER",
+                accepted: true,
+              },
+              {
+                userId: user.id,
+                teamId: orgId,
+                role: "MEMBER",
+                accepted: true,
+              },
+            ];
+          })
+          .flat(),
+      ],
+      skipDuplicates: true,
```

Comment:
I'm really liking `skipDuplicates` this will create memberships for all users that are new to the team but will also skip over creating memberships if the user is already a part of the org.